wayland-client/src/event_queue.rs

```diff
@@ -496,53 +517,94 @@ impl<State> EventQueue<State> {
             if let Err(e) = self.conn.backend.dispatch_inner_queue() {
                 return task::Poll::Ready(Err(e.into()));
             }
-            match Pin::new(&mut self.rx).poll_next(cx) {
-                task::Poll::Pending => return task::Poll::Pending,
-                task::Poll::Ready(None) => {
-                    // We never close the channel, and we hold a valid sender in self.handle.tx, so
-                    // our event stream will never reach an end.
-                    unreachable!("Got end of stream while holding a valid sender");
-                }
-                task::Poll::Ready(Some(QueueEvent(cb, msg, odata))) => {
-                    cb(&self.conn, msg, data, odata, &self.handle)?
-                }
+            let mut lock = self.handle.inner.lock().unwrap();
+            if lock.freeze_count != 0 {
+                lock.waker = Some(cx.waker().clone());
+                return task::Poll::Pending;
             }
+            let QueueEvent(cb, msg, odata) = if let Some(elt) = lock.queue.pop_front() {
+                elt
+            } else {
+                lock.waker = Some(cx.waker().clone());
+                return task::Poll::Pending;
+            };
+            drop(lock);
+            cb(&self.conn, msg, data, odata, &self.handle)?
         }
     }
 }
 
+struct DispatchWaker {
+    cond: Condvar,
+}
+
+impl task::Wake for DispatchWaker {
+    fn wake(self: Arc<Self>) {
+        self.cond.notify_all()
+    }
+}
+
 /// A handle representing an [`EventQueue`], used to assign objects upon creation.
 pub struct QueueHandle<State> {
-    tx: UnboundedSender<QueueEvent<State>>,
+    inner: Arc<Mutex<EventQueueInner<State>>>,
+}
+
+/// A handle that temporarily pauses event processing on an [`EventQueue`].
+#[derive(Debug)]
+pub struct FreezeHandle<'a, State> {
```

Comment:
Is there a better name that could be bikeshedded for this type?

Reply:
I'm thinking `QueueFreezeGuard` instead, illustrating that it behaves more like a guard (releasing a resource on drop).